index.js

```diff
@@ -253,59 +268,69 @@ function fastifyView (fastify, opts, next) {
+    if (!localOptions) {
+      localOptions = globalOptions
+    }
+    if (type === 'ejs' && ((localOptions.async ?? globalOptions.async) || cachedPage instanceof Promise)) {
+      cachedPage.then(html => {
+        if (useHtmlMinification(globalOptions, requestedPath)) {
+          html = globalOptions.useHtmlMinifier.minify(html, globalOptions.htmlMinifierOptions || {})
+        }
+        that.send(html)
+      }).catch(err => that.send(err))
+      return
+    }
+    if (useHtmlMinification(globalOptions, requestedPath)) {
+      cachedPage = globalOptions.useHtmlMinifier.minify(cachedPage, globalOptions.htmlMinifierOptions || {})
+    }
+    that.send(cachedPage)
+  }
 
-      let compiledPage
-      try {
-        if ((type === 'ejs') && viewExt && !globalOptions.includer) {
-          globalOptions.includer = (originalPath, parsedPath) => {
-            return {
-              filename: parsedPath || join(templatesDir, originalPath + '.' + viewExt)
-            }
+  function readCallbackParser (that, page, html, localOptions) {
+    let compiledPage
+    try {
+      if ((type === 'ejs') && viewExt && !globalOptions.includer) {
+        globalOptions.includer = (originalPath, parsedPath) => {
+          return {
+            filename: parsedPath || join(templatesDir, originalPath + '.' + viewExt)
           }
         }
-        globalOptions.filename = join(templatesDir, page)
-        if (localOptions) {
-          for (const key in globalOptions) {
-            if (!Object.prototype.hasOwnProperty.call(localOptions, key)) localOptions[key] = globalOptions[key]
-          }
-        } else localOptions = globalOptions
-        compiledPage = engine.compile(html, localOptions)
-      } catch (error) {
-        that.send(error)
-        return
       }
-      lru.set(page, compiledPage)
+      if (localOptions) {
+        for (const key in globalOptions) {
+          if (!Object.prototype.hasOwnProperty.call(localOptions, key)) localOptions[key] = globalOptions[key]
+        }
+      } else localOptions = globalOptions
+      compiledPage = engine.compile(html, localOptions)
+    } catch (error) {
+      that.send(error)
+      return
+    }
+    lru.set(page, compiledPage)
+    return compiledPage
+  }
 
-      if (!that.getHeader('content-type')) {
-        that.header('Content-Type', 'text/html; charset=' + charset)
-      }
-      let cachedPage
-      try {
-        cachedPage = lru.get(page)(data)
-      } catch (error) {
-        cachedPage = error
-      }
-      if (type === 'ejs' && (localOptions.async ?? globalOptions.async)) {
-        cachedPage.then(html => {
-          if (useHtmlMinification(globalOptions, requestedPath)) {
-            html = globalOptions.useHtmlMinifier.minify(html, globalOptions.htmlMinifierOptions || {})
-          }
-          that.send(html)
-        }).catch(err => that.send(err))
+  function readCallback (that, page, data, localOptions) {
+    return function _readCallback (err, html) {
```

Comment:
@mcollina don't nested functions hurt performance? Wouldn't it be better to inline the callback and just access the properties from there?

I guess it's fine since a file will be read only once in production

Reply:
I guess if you want to change that, it should be in another PR
https://github.com/fastify/point-of-view/blob/v8.1.0/index.js#L256-L257

Reply:
Looks fine even in non-production:

Master:
```shell
Running 10s test @ http://127.0.0.1:3000/
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 0 ms  │ 0 ms │ 0.01 ms │ 0.06 ms │ 14 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬──────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg      │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Req/Sec   │ 46495   │ 46495   │ 61919   │ 62751   │ 60536.73 │ 4487.85 │ 46481   │
├───────────┼─────────┼─────────┼─────────┼─────────┼──────────┼─────────┼─────────┤
│ Bytes/Sec │ 12.5 MB │ 12.5 MB │ 16.6 MB │ 16.8 MB │ 16.2 MB  │ 1.2 MB  │ 12.5 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴──────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

666k requests in 11.01s, 178 MB read
```

PR:
```shell
Running 10s test @ http://127.0.0.1:3000/
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 0 ms  │ 0 ms │ 0.01 ms │ 0.06 ms │ 14 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬───────┬──────────┬─────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5% │ Avg      │ Stdev   │ Min     │
├───────────┼─────────┼─────────┼─────────┼───────┼──────────┼─────────┼─────────┤
│ Req/Sec   │ 57439   │ 57439   │ 62975   │ 63519 │ 62244.37 │ 1731.73 │ 57412   │
├───────────┼─────────┼─────────┼─────────┼───────┼──────────┼─────────┼─────────┤
│ Bytes/Sec │ 15.4 MB │ 15.4 MB │ 16.9 MB │ 17 MB │ 16.7 MB  │ 463 kB  │ 15.4 MB │
└───────────┴─────────┴─────────┴─────────┴───────┴──────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

685k requests in 11.01s, 183 MB read
```

Reply:
Wow, I believe it is really faster